### PR TITLE
Introduce Memo.Run.Fdecl

### LINKED ISSUE #2941: Polling mode seems broken
(probably related to #2834)
This seems to break the polling mode. 
I'm getting:

```
("Fdecl.set: already set",
old = ...
new = ... <same as old>
})
Backtrace:
Raised at file "src/stdune/code_error.ml", line 9, characters 30-62
Called from file "src/dune/main.ml", line 39, characters 2-26
Called from file "src/dune/main.ml", line 44, characters 12-38
Called from file "bin/import.ml", line 75, characters 21-42
Called from file "src/fiber/fiber.ml", line 114, characters 10-15
```

### PR DESCRIPTION
**Pull request by rgrinberg**
Same as Fdecl, but is reset after every run. With this new Fdecl, it should be easier to make sure this doesn't break going forward

Fix #2941

### CODE REVIEW

**Comment by aalekseyev** [comment]
This module seems tricky to use correctly. I wonder if the call sites should be responsible for some of this.

In particular, this part:
```
(fun () ->
     let (_ : Run.t) = current_run () in
        Fdecl.create to_dyn)
```

would be OK to have at the call site, it seems? (because both fdecl and current_run are things you have to keep in mind to reason about `Memo.Run.Fdecl`)

And if we do that, then the rest of the module is exactly `Memo.lazy`. 
With that in mind, maybe we should just let the call site do all 3 things (Memo.lazy, current_run and Fdecl.create) and we don't need this module?

Maybe there's a way to think about the module that doesn't require mentally disassembling it. I'm happy to hear any ideas on this front.

**Comment by rgrinberg** [comment]
The issue with doing all this work on the call site is that it’s very easy to forget. That much is clear given how many bugs we’ve had related to this.

That being said, I fully agree that this isn’t a fully baked abstraction and all that it accomplishes is sharing an implementation.

Perhaps demoting it from the Memo library to dune itself is sufficient? Otherwise we’ll simply move the extra mental effort to the call sites.

**Comment by aalekseyev** [comment]
All right, I think it's fine to keep it as long as we're careful when we use it. I don't think it should be moved to Dune from Memo. I think that achieves nothing except making it harder to find.

File: src/memo/memo.ml [hunk 1]
```diff
@@ -800,3 +799,28 @@ module Implicit_output = Implicit_output
 module Store = Store_intf
 
 let on_already_reported f = on_already_reported := f
+
+module Run = struct
+  module Fdecl = struct
+    type nonrec 'a t = { memo : (unit, 'a Fdecl.t, unit -> 'a Fdecl.t) t }
+
+    let create to_dyn =
+      let memo =
```
Comment (aalekseyev) [inline src/memo/memo.ml]: I think `Memo.lazy_` should be used here though.

**Comment by rgrinberg** [comment]
> I don't think it should be moved to Dune from Memo. I think that achieves nothing except making it harder to find.

Well it would make it possible to keep the implementation private. Provided that we define all these run vars in one place.

**Comment by aalekseyev** [comment]
Ah, in that case that could indeed be worth it.

**Pull request merged**